src/libxtp/sternheimerw.cc

```diff
@@ -0,0 +1,404 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ */
+
+#include <ctime>
+#include <eigen3/Eigen/src/Core/Block.h>
+#include <eigen3/Eigen/src/Core/Matrix.h>
+#include <eigen3/Eigen/src/Core/PlainObjectBase.h>
+#include <eigen3/Eigen/src/Eigenvalues/GeneralizedEigenSolver.h>
+#include <eigen3/Eigen/src/Eigenvalues/SelfAdjointEigenSolver.h>
+#include <eigen3/Eigen/src/IterativeLinearSolvers/BiCGSTAB.h>
+#include <eigen3/Eigen/src/IterativeLinearSolvers/ConjugateGradient.h>
+#include <eigen3/Eigen/src/IterativeLinearSolvers/IterativeSolverBase.h>
+#include <fstream>
+#include <math.h>
+#include <votca/tools/property.h>
+#include <votca/xtp/aobasis.h>
+#include <votca/xtp/aomatrix.h>
+#include <votca/xtp/logger.h>
+#include <votca/xtp/multishift.h>
+#include <votca/xtp/orbitals.h>
+#include <votca/xtp/padeapprox.h>
+#include <votca/xtp/sternheimerw.h>
+
+#include "votca/xtp/numerical_integrations.h"
+
+namespace votca {
+namespace xtp {
+
+void SternheimerW::Setup() {
+
+  this->_num_occ_lvls = _orbitals.getNumberOfAlphaElectrons();
+  this->_basis_size = _orbitals.getBasisSetSize();
+  this->_H = Hamiltonian();
+  this->_p = _orbitals.DensityMatrixGroundState();
+  this->_S = OverlapMatrix();
+  this->_mo_coefficients = _orbitals.MOCoefficients();
+  this->_mo_energies = _orbitals.MOEnergies();
+}
+
+void SternheimerW::initializeMultishift(int basis_size) {
+
+  Multishift multi;
+  this->_multishift = multi;
+
+  _multishift.setBasisSize(basis_size);
+}
+void SternheimerW::initializePade(int size) {
+
+  PadeApprox pade;
+  this->_pade = pade;
+
+  _pade.initialize(size);
+}
+
+Eigen::MatrixXd SternheimerW::OverlapMatrix() {
+
+  AOBasis basis = _orbitals.SetupDftBasis();
+  AOOverlap overlap;
+  overlap.Fill(basis);
+  return overlap.Matrix();
+}
+
+Eigen::MatrixXd SternheimerW::DensityMatrix() {
+
+  return _orbitals.DensityMatrixGroundState();
+}
+
+Eigen::MatrixXd SternheimerW::Hamiltonian() {
+
+  const Eigen::MatrixXd& mo_coefficients = _orbitals.MOCoefficients();
+  const Eigen::MatrixXd& mo_energies = _orbitals.MOEnergies().asDiagonal();
+  const Eigen::MatrixXd overlap = OverlapMatrix();
+  Eigen::MatrixXd H = overlap * mo_coefficients * mo_energies *
+                      mo_coefficients.transpose() * overlap;
+
+  return H;
+}
+Eigen::MatrixXcd SternheimerW::CoulombMatrix(Eigen::Vector3d gridpoint) {
+
+  AOBasis basis = _orbitals.SetupDftBasis();
+  AOESP aoesp;
+  aoesp.setPosition(gridpoint);
+  aoesp.Fill(basis);
+  return aoesp.Matrix();
+}
+
+Eigen::MatrixXcd SternheimerW::SternheimerLHS(Eigen::MatrixXcd hamiltonian,
+                                              Eigen::MatrixXcd overlap,
+                                              double eps,
+                                              std::complex<double> omega,
+                                              bool pm) {
+
+  Eigen::MatrixXcd S;
+  // distinguish between +w and -w
+  if (pm == true) {
```

Comment:
see earlier. This code is to a good part a copy of the sternheimer.cc don't be afraid to delete code and files as long as it is all under version control